File: scripts/build-blog.js
#!/usr/bin/env node
/**
 * build-blog.js
 * Compile blog/posts/*.md → blog/*.html and regenerate blog/index.html
 *
 * Frontmatter (YAML-style, between --- delimiters):
 *   title: My Post Title
 *   date: 2026-07-12
 *   description: A short summary shown in the index.
 *   slug: my-post-title        # optional; defaults to filename without .md
 *   draft: true                # optional; skip this post entirely
 *
 * Usage:
 *   node scripts/build-blog.js
 */

import { readFileSync, writeFileSync, readdirSync, mkdirSync, existsSync } from 'fs';
import { join, basename, dirname } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const ROOT = join(__dirname, '..');
const POSTS_DIR = join(ROOT, 'blog', 'posts');
const BLOG_DIR = join(ROOT, 'blog');

// ---------------------------------------------------------------------------
// Minimal Markdown → HTML converter (no deps)
// ---------------------------------------------------------------------------
function mdToHtml(md) {
  let html = md;

  // Fenced code blocks  ```lang\n...\n```
  html = html.replace(/```(\w*)\n([\s\S]*?)```/g, (_, lang, code) => {
    const escaped = escapeHtml(code.trimEnd());
    const cls = lang ? ` class="language-${lang}"` : '';
    return `<pre><code${cls}>${escaped}</code></pre>`;
  });

  // Inline code `...`
  html = html.replace(/`([^`\n]+)`/g, (_, c) => `<code>${escapeHtml(c)}</code>`);

  // Images  ![alt](src)
  html = html.replace(/!\[([^\]]*)\]\(([^)]+)\)/g, '<img src="$2" alt="$1">');

  // Links  [text](url)
  html = html.replace(/\[([^\]]+)\]\(([^)]+)\)/g, '<a href="$2">$1</a>');

  // Headings (must come before bold/italic)
  html = html.replace(/^#{6}\s+(.+)$/gm, '<h6>$1</h6>');
  html = html.replace(/^#{5}\s+(.+)$/gm, '<h5>$1</h5>');
  html = html.replace(/^#{4}\s+(.+)$/gm, '<h4>$1</h4>');
  html = html.replace(/^###\s+(.+)$/gm, '<h3>$1</h3>');
  html = html.replace(/^##\s+(.+)$/gm, '<h2>$1</h2>');
  html = html.replace(/^#\s+(.+)$/gm, '<h1>$1</h1>');

  // Horizontal rule
  html = html.replace(/^[-*_]{3,}\s*$/gm, '<hr>');

  // Blockquotes
  html = html.replace(/^>\s+(.+)$/gm, '<blockquote><p>$1</p></blockquote>');

  // Bold + italic ***text***
  html = html.replace(/\*{3}(.+?)\*{3}/g, '<strong><em>$1</em></strong>');
  // Bold **text** or __text__
  html = html.replace(/\*\*(.+?)\*\*/g, '<strong>$1</strong>');
  html = html.replace(/__(.+?)__/g, '<strong>$1</strong>');
  // Italic *text* or _text_
  html = html.replace(/\*([^*\n]+)\*/g, '<em>$1</em>');
  html = html.replace(/_([^_\n]+)_/g, '<em>$1</em>');

  // Unordered lists (-, *, +)
  html = html.replace(/((?:^[-*+] .+\n?)+)/gm, (block) => {
    const items = block.trim().split('\n').map(l => `  <li>${l.replace(/^[-*+] /, '').trim()}</li>`).join('\n');
    return `<ul>\n${items}\n</ul>\n`;
  });

  // Ordered lists
  html = html.replace(/((?:^\d+\. .+\n?)+)/gm, (block) => {
    const items = block.trim().split('\n').map(l => `  <li>${l.replace(/^\d+\. /, '').trim()}</li>`).join('\n');
    return `<ol>\n${items}\n</ol>\n`;
  });

  // Paragraphs: wrap consecutive non-empty, non-block lines
  html = html.replace(/^(?!<[a-z]|$)(.+)$/gm, (line) => {
    // Don't double-wrap already-tagged lines
    if (/^<(h[1-6]|ul|ol|li|blockquote|pre|hr|img)/.test(line.trim())) return line;
    return `<p>${line}</p>`;
  });

  // Clean up blank lines
  html = html.replace(/\n{3,}/g, '\n\n');

  return html.trim();
}

function escapeHtml(str) {
  return str
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;');
}

// ---------------------------------------------------------------------------
// Frontmatter parser
// ---------------------------------------------------------------------------
function parseFrontmatter(raw) {
  const match = raw.match(/^---\n([\s\S]*?)\n---\n([\s\S]*)$/);
  if (!match) return { meta: {}, body: raw };

  const meta = {};
  for (const line of match[1].split('\n')) {
    const idx = line.indexOf(':');
    if (idx === -1) continue;
    const key = line.slice(0, idx).trim();
    const val = line.slice(idx + 1).trim().replace(/^['"]|['"]$/g, '');
    meta[key] = val;
  }
  return { meta, body: match[2] };
}

// ---------------------------------------------------------------------------
// HTML template for a single post
// ---------------------------------------------------------------------------
function renderPost({ title, date, description, slug, body }) {
  const safeTitle = escapeHtml(title);
  const safeDesc = escapeHtml(description || '');
  return `<!doctype html><html lang="en"><head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<title>${safeTitle} | Eric Spencer</title>
<meta name="description" content="${safeDesc}">
<meta name="author" content="Eric Spencer">
<meta name="robots" content="index, follow">
<link rel="canonical" href="https://ericspencer.us/blog/${slug}.html">
<meta property="og:type" content="article">
<meta property="og:title" content="${safeTitle} | Eric Spencer">
<meta property="og:description" content="${safeDesc}">
<meta property="og:url" content="https://ericspencer.us/blog/${slug}.html">
<meta property="og:site_name" content="Eric Spencer">
<link rel="preconnect" href="https://fonts.googleapis.com">
<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
<link rel="stylesheet" href="https://fonts.googleapis.com/css2?family=IBM+Plex+Mono:wght@400;500&family=Plus+Jakarta+Sans:wght@300;400;500;600;700&display=swap">
<style>
:root{--paper:oklch(0.978 0.006 80);--ink:oklch(0.11 0.015 60);--accent:oklch(0.30 0.13 22);--dim:oklch(0.52 0.01 70);--rule:oklch(0.87 0.005 80)}
*{box-sizing:border-box}html{scroll-behavior:smooth}
body{margin:0;background:var(--paper);color:var(--ink);font-family:"Plus Jakarta Sans",-apple-system,BlinkMacSystemFont,sans-serif;font-size:16px;line-height:1.7;-webkit-font-smoothing:antialiased}
.wrap{max-width:680px;margin:0 auto;padding:56px 24px 100px}
nav.top{font-size:14px;font-weight:500;color:var(--dim);margin:0 0 32px}
nav.top a{color:inherit;text-decoration:none;border:0;padding:0 2px}
nav.top a:hover{color:var(--ink)}
hr{border:0;border-top:1px solid var(--rule);margin:28px 0}
a{color:var(--accent);text-decoration:none}
a:hover{text-decoration:underline;text-decoration-thickness:1px;text-underline-offset:3px}
h1{font-size:clamp(24px,4vw,36px);font-weight:700;letter-spacing:-0.025em;line-height:1.1;margin:0 0 8px}
.meta{font-family:"IBM Plex Mono",monospace;font-size:12px;color:var(--dim);margin:0 0 32px}
h2{font-size:12px;font-weight:600;text-transform:uppercase;letter-spacing:0.07em;color:var(--dim);margin:40px 0 12px}
h3{font-size:18px;font-weight:600;letter-spacing:-0.01em;margin:28px 0 6px}
p{margin:14px 0;font-size:16px;line-height:1.75}
code,kbd,pre{font-family:"IBM Plex Mono",ui-monospace,monospace}
code{background:oklch(0.94 0.005 80);padding:1px 5px;border-radius:4px;font-size:13px}
pre{background:oklch(0.12 0.01 240);color:oklch(0.85 0.02 80);padding:14px 18px;border-radius:6px;overflow-x:auto;font-size:13px;line-height:1.6;margin:16px 0}
pre code{background:none;padding:0;color:inherit}
blockquote{margin:16px 0;padding:8px 18px;border-left:2px solid var(--rule);color:var(--dim);font-style:italic}
ul,ol{margin:12px 0;padding-left:24px}li{margin:5px 0}
img{max-width:100%;border-radius:6px;margin:16px 0}
footer{margin-top:60px;border-top:1px solid var(--rule);padding-top:16px;font-size:13px;color:var(--dim);display:flex;justify-content:space-between;flex-wrap:wrap;gap:8px}
@media(max-width:560px){.wrap{padding:36px 18px 80px}}
</style></head><body><div class="wrap">

<nav class="top"><a href="/">index</a> &nbsp;·&nbsp; <a href="/blog/">&larr; blog</a></nav>

<h1>${safeTitle}</h1>
<div class="meta">${date || ''}</div>
<hr>

${body}

<footer>
<span>&copy; ${new Date().getFullYear()} Eric Spencer &middot; Chicago, IL</span>
<span><a href="mailto:eric@ericspencer.us">eric@ericspencer.us</a></span>
</footer>
</div></body></html>`;
}

// ---------------------------------------------------------------------------
// Regenerate blog/index.html post list
// ---------------------------------------------------------------------------
function rebuildIndex(posts) {
  const indexPath = join(BLOG_DIR, 'index.html');
  let html = readFileSync(indexPath, 'utf8');

  // Build rows HTML (newest first)
  const sorted = [...posts].sort((a, b) => (b.date || '').localeCompare(a.date || ''));
  let rows;
  if (sorted.length === 0) {
    rows = `\n<p style="color:var(--dim);font-size:14px;font-style:italic">No posts yet.</p>\n`;
  } else {
    rows = '\n' + sorted.map(p => {
      const desc = p.description ? `\n<div class="post-desc">${escapeHtml(p.description)}</div>` : '';
      return `<div class="post-row"><div class="post-d">${p.date || ''}</div><div class="post-body"><div class="post-title"><a href="/blog/${p.slug}.html">${escapeHtml(p.title)}</a></div>${desc}</div></div>`;
    }).join('\n') + '\n';
  }

  // Replace everything between <!-- POSTS:START --> and <!-- POSTS:END --> markers
  if (html.includes('<!-- POSTS:START -->')) {
    html = html.replace(/<!-- POSTS:START -->[\s\S]*?<!-- POSTS:END -->/, `<!-- POSTS:START -->${rows}<!-- POSTS:END -->`);
  } else {
    // First run: insert markers around whatever placeholder is there now
    html = html.replace(
      /<h2>Posts<\/h2>\n[\s\S]*?(?=\n<footer)/,
      `<h2>Posts</h2>\n<!-- POSTS:START -->${rows}<!-- POSTS:END -->`
    );
  }

  writeFileSync(indexPath, html, 'utf8');
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------
if (!existsSync(POSTS_DIR)) {
  mkdirSync(POSTS_DIR, { recursive: true });
  console.log('Created blog/posts/ — drop .md files there.');
}

const mdFiles = readdirSync(POSTS_DIR).filter(f => f.endsWith('.md') && !f.startsWith('_'));
if (mdFiles.length === 0) {
  console.log('No .md files found in blog/posts/. Nothing to build.');
  process.exit(0);
}

const built = [];

for (const file of mdFiles) {
  const raw = readFileSync(join(POSTS_DIR, file), 'utf8');
  const { meta, body } = parseFrontmatter(raw);

  if (meta.draft === 'true') {
    console.log(`  skip (draft)  ${file}`);
    continue;
  }

  const slug = meta.slug || basename(file, '.md');
  const title = meta.title || slug;
  const date = meta.date || '';
  const description = meta.description || '';

  const htmlBody = mdToHtml(body);
  const page = renderPost({ title, date, description, slug, body: htmlBody });

  const outPath = join(BLOG_DIR, `${slug}.html`);
  writeFileSync(outPath, page, 'utf8');
  console.log(`  built  blog/${slug}.html`);

  built.push({ slug, title, date, description });
}

rebuildIndex(built);
console.log(`\nDone. ${built.length} post(s) compiled. blog/index.html updated.`);

File: scripts/build_blog.py
#!/usr/bin/env python3
"""Build blog/*.html and blog/index.html from content/blog/*.md."""

from __future__ import annotations

import html
import re
from dataclasses import dataclass
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
SOURCE = ROOT / "content" / "blog"
OUTPUT = ROOT / "blog"


@dataclass(frozen=True)
class Post:
    slug: str
    title: str
    date: str
    description: str
    body: str


def parse_front_matter(path: Path) -> Post | None:
    raw = path.read_text(encoding="utf-8")
    match = re.match(r"\A---\s*\n(.*?)\n---\s*\n(.*)\Z", raw, re.DOTALL)
    if not match:
        raise ValueError(f"{path}: expected YAML front matter between --- lines")

    fields = {}
    for line in match.group(1).splitlines():
        if not line.strip() or line.lstrip().startswith("#"):
            continue
        key, separator, value = line.partition(":")
        if not separator:
            raise ValueError(f"{path}: invalid front matter line: {line}")
        fields[key.strip()] = value.strip().strip("\"'")

    if fields.get("published", "true").lower() in {"false", "no", "draft"}:
        return None
    for field in ("title", "date", "description"):
        if not fields.get(field):
            raise ValueError(f"{path}: missing front matter field '{field}'")

    slug = fields.get("slug", path.stem)
    if not re.fullmatch(r"[a-z0-9]+(?:-[a-z0-9]+)*", slug):
        raise ValueError(f"{path}: slug must contain lowercase letters, numbers, and hyphens")
    if not re.fullmatch(r"\d{4}-\d{2}-\d{2}", fields["date"]):
        raise ValueError(f"{path}: date must use YYYY-MM-DD")

    return Post(slug, fields["title"], fields["date"], fields["description"], match.group(2).strip())


def load_posts() -> list[Post]:
    if not SOURCE.exists():
        return []
    posts = [post for path in sorted(SOURCE.glob("*.md")) if (post := parse_front_matter(path))]
    return sorted(posts, key=lambda post: post.date, reverse=True)


def inline_markdown(value: str) -> str:
    escaped = html.escape(value, quote=False)
    placeholders: list[str] = []

    def protect(fragment: str) -> str:
        placeholders.append(fragment)
        return f"\x00{len(placeholders) - 1}\x00"

    escaped = re.sub(
        r"!\[([^\]]*)\]\(([^)\s]+)(?:\s+['\"]([^'\"]*)['\"])?\)",
        lambda match: protect(
            f'<img src="{html.escape(match.group(2), quote=True)}" '
            f'alt="{html.escape(match.group(1), quote=True)}">'
        ),
        escaped,
    )
    escaped = re.sub(
        r"\[([^\]]+)\]\(([^)\s]+)\)",
        lambda match: protect(
            f'<a href="{html.escape(match.group(2), quote=True)}">{match.group(1)}</a>'
        ),
        escaped,
    )
    escaped = re.sub(r"`([^`]+)`", r"<code>\1</code>", escaped)
    escaped = re.sub(r"\*\*(.+?)\*\*|__(.+?)__", lambda match: f"<strong>{match.group(1) or match.group(2)}</strong>", escaped)
    escaped = re.sub(r"(?<!\*)\*([^*]+)\*(?!\*)|(?<!_)_([^_]+)_(?!_)", lambda match: f"<em>{match.group(1) or match.group(2)}</em>", escaped)
    escaped = re.sub(
        r"(?<![\w\"=])(https?://[^\s<]+)",
        lambda match: protect(f'<a href="{match.group(1)}" target="_blank" rel="noopener">{match.group(1)}</a>'),
        escaped,
    )
    for index, fragment in enumerate(placeholders):
        escaped = escaped.replace(f"\x00{index}\x00", fragment)
    return escaped


def markdown_to_html(markdown: str) -> str:
    output: list[str] = []
    paragraph: list[str] = []
    list_type: str | None = None
    in_code = False
    code_lines: list[str] = []

    def flush_paragraph() -> None:
        if paragraph:
            output.append(f"<p>{inline_markdown(' '.join(paragraph))}</p>")
            paragraph.clear()

    def close_list() -> None:
        nonlocal list_type
        if list_type:
            output.append(f"</{list_type}>")
            list_type = None

    for line in markdown.splitlines():
        if line.startswith("```"):
            if in_code:
                output.append(f"<pre><code>{html.escape(chr(10).join(code_lines))}</code></pre>")
                code_lines.clear()
            else:
                flush_paragraph()
                close_list()
            in_code = not in_code
            continue
        if in_code:
            code_lines.append(line)
            continue

        stripped = line.strip()
        if not stripped:
            flush_paragraph()
            close_list()
            continue
        heading = re.match(r"^(#{1,3})\s+(.+)$", stripped)
        if heading:
            flush_paragraph()
            close_list()
            level = len(heading.group(1)) + 1
            output.append(f"<h{level}>{inline_markdown(heading.group(2))}</h{level}>")
            continue
        if re.fullmatch(r"[-*_]{3,}", stripped):
            flush_paragraph()
            close_list()
            output.append("<hr>")
            continue
        quote = re.match(r"^>\s?(.*)$", stripped)
        if quote:
            flush_paragraph()
            close_list()
            output.append(f"<blockquote>{inline_markdown(quote.group(1))}</blockquote>")
            continue
        item = re.match(r"^([-*+] |\d+\. )(.+)$", stripped)
        if item:
            flush_paragraph()
            requested_type = "ol" if item.group(1)[0].isdigit() else "ul"
            if list_type != requested_type:
                close_list()
                output.append(f"<{requested_type}>")
                list_type = requested_type
            output.append(f"<li>{inline_markdown(item.group(2))}</li>")
            continue
        close_list()
        paragraph.append(stripped)

    if in_code:
        raise ValueError("unterminated fenced code block")
    flush_paragraph()
    close_list()
    return "\n".join(output)


def render_post(post: Post) -> str:
    template = (OUTPUT / "_template.html").read_text(encoding="utf-8")
    replacements = {
        "POST TITLE": html.escape(post.title),
        "POST DESCRIPTION": html.escape(post.description, quote=True),
        "YOUR-SLUG": post.slug,
        "YYYY-MM-DD": post.date,
        "<!-- Write your post content here -->\n<p>...</p>": markdown_to_html(post.body),
    }
    for old, new in replacements.items():
        template = template.replace(old, new)
    return template


def render_index(posts: list[Post]) -> str:
    path = OUTPUT / "index.html"
    index = path.read_text(encoding="utf-8")
    rows = "\n".join(
        f'<div class="post-row"><span class="post-d">{post.date}</span>'
        f'<div class="post-body"><div class="post-title"><a href="/blog/{post.slug}.html">'
        f"{html.escape(post.title)}</a></div><p class=\"post-desc\">"
        f"{html.escape(post.description)}</p></div></div>"
        for post in posts
    ) or '<p style="color:var(--dim);font-size:14px;font-style:italic">No posts yet.</p>'
    start = "<!-- POSTS_START -->"
    end = "<!-- POSTS_END -->"
    if start not in index or end not in index:
        raise ValueError("blog/index.html: missing POSTS_START/POSTS_END markers")
    before, remainder = index.split(start, 1)
    _, after = remainder.split(end, 1)
    return f"{before}{start}\n{rows}\n\n{end}{after}"


def prune_orphans(posts: list[Post]) -> list[str]:
    """Delete post pages whose source is gone or has gone back to draft.

    Flipping a post to `published: false` used to leave its HTML behind, so an
    unpublished draft stayed reachable by direct link and stayed in the sitemap.
    """
    keep = {f"{post.slug}.html" for post in posts} | {"index.html", "_template.html"}
    removed = []
    for path in sorted(OUTPUT.glob("*.html")):
        if path.name not in keep:
            path.unlink()
            removed.append(path.name)
    return removed


def main() -> None:
    posts = load_posts()
    OUTPUT.mkdir(exist_ok=True)
    for post in posts:
        (OUTPUT / f"{post.slug}.html").write_text(render_post(post), encoding="utf-8")
    (OUTPUT / "index.html").write_text(render_index(posts), encoding="utf-8")
    removed = prune_orphans(posts)
    print(f"Built {len(posts)} blog post(s)"
          + (f"; removed {', '.join(removed)}" if removed else ""))


if __name__ == "__main__":
    main()